keps/sig-auth/1205-bound-service-account-tokens/README.md

```diff
@@ -265,53 +258,161 @@ service account token on behalf of pods running on that node. The
-  1. Add annotation to audit events for legacy and stale tokens including necessary 
-  information to locate problematic client.
- 
-This functionality can be implemented entirely in the kube-apiserver and does not 
-require cooperation by the kubelet beyond the standard functionality implemented 
-as part of [Service Account Token Volumes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/svcacct-token-volume-source.md).
+1.  Compare the `kubernetes.io/warnafter` field with current time. If current
+    time is after `kubernetes.io/warnafter` field, it implies calling client is
+    not reloading token regularly.
+1.  Expose metrics to monitor number of legacy and stale token used.
+1.  Add annotation to audit events for legacy and stale tokens including
+    necessary information to locate problematic client.
 
-We will direct cluster administrators to adopt bound service account tokens and 
-turn on the command line flag to enable extended token. Fix clients if monitoring or 
-audit report that stale or legacy token are in use. The operator can turn off 
-the command line flag after no stale token is being used.
+### Graduation Criteria
 
+#### Alpha->Beta
 
+All known migration frictions have been fixed.
 
-### Graduation Criteria
+- PodSecurityPolicies that allow secrets but not projected volumes will
+  handicap the provision of token volumes.
+  - Fixed in https://github.com/kubernetes/kubernetes/pull/92006
+- In-cluster clients that don’t reload service account tokens will start
+  failing an hour after deployment.
+  - Fixed in https://github.com/kubernetes/kubernetes/issues/68164
+- Pods running as non root may not access the service account token.
+  - Fixed in https://github.com/kubernetes/kubernetes/pull/89193
 
 #### Beta -> GA Graduation
 
-- TBD
+New `ServiceAccount` admission controller WAI in Beta for >= 1 minor without
+significant issues.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  - Feature gate name: `BoundServiceAccountVolume`
+  - Components depending on the feature gate: kubelet, kube-apiserver
+  - Will enabling / disabling the feature require downtime of the control
+    plane? yes, need to restart kube apiserver.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? yes, need to restart kubelet.
+
+- **Does enabling the feature change any default behavior?** yes, pods'
+  service account tokens will not be long-lived and are not stored as Secrets
+  any more.
+
+- **Can the feature be disabled once it has been enabled (i.e. can we roll
+  back the enablement)?** yes.
+
+- **What happens if we reenable the feature if it was previously rolled
+  back?** the same as the first enablement.
+
+- **Are there any tests for feature enablement/disablement?** yes, unit tests
+  will cover this.
```

Comment:
unit tests cover specific functions' behavior with the feature enabled/disabled, but do not test scenarios like:
1. deploy pod A with feature disabled
2. enable feature, ensure pod A keeps working
3. deploy pod B with feature enabled
4. disable feature, ensure pod A and B keep working
5. enable feature, ensure pod A and B keep working

Reply:
did you mean adding integration test to cover these cases? if so, can i say pod working == pod object exists in kube-apiserver? i am not sure how to implement an e2e test for cover the these cases because it requires restart of kube-apiserver and kube-controller-manager.

Reply:
> did you mean adding integration test to cover these cases? if so, can i say pod working == pod object exists in kube-apiserver?

not really... this is a transition that would impact all workloads, so we really need to exercise the scenario in a realistic way.

> i am not sure how to implement an e2e test for cover the these cases because it requires restart of kube-apiserver and kube-controller-manager.

yeah... our upgrade or config change test story is not great at the moment. 

we do have an upgrade test framework (see test/e2e/upgrades) where we could at least exercise steps 1-3 on the release boundary where the feature becomes enabled by default (assuming those test jobs provide the token signing args to the kube-apiserver, which they should start doing if they are not already).

something along the lines of https://github.com/kubernetes/kubernetes/issues/78904 would be needed to be sure the particular pod instances created before the upgrade were still functional after the upgrade.

Reply:
drafted up an upgrade test here https://github.com/zshihang/kubernetes/commit/4c7c94b0339b24d98fa91531a5249e48d4c846e8
but i failed to run any upgrade test locally. the first issue came up due to the sha1 is not published in ci/latest builds while gce startup scripts rely on them. after fixing the hash issue, ran into a coredns issue. get the upgrade test working locally right now fixing those issues seems a snowballing.

talked to @BenTheElder yesterday, the upgrade tests are not monitored so even get this new test working locally and merged, we still have a gap in exercising the steps 1-3. however, i have tested the steps 1-3 manually in gke where pod keeps working as expected, which gives me confidence that we should not block the beta graduation by the ongoing work for kubetest2 and a better upgrade testing framework.

Reply:
Before we automatically enable this feature in existing environments on upgrade, we need tests that ensure that path will function properly. It's very unfortunate we don't have a ready/functional framework for tests like that, but it is critical we test this in an automated recurring way.

Reply:
opened https://github.com/kubernetes/kubernetes/issues/94487 to track the issue that we currently have no upgrade test signal

Reply:
finally get the [upgrade test](https://github.com/kubernetes/kubernetes/pull/94835) passing locally. next step would be fixing the hash issue so that the version ci/latest points to will work and then set up ci job to run it periodically.

we can merge this PR if we agree on the timeline.
1.20: TokenRequest GA, BoundServiceAccountToken Beta
1.21+: BoundServiceAccountToken GA

Reply:
indicate that unit tests already exist for individual components, and an e2e upgrade test is part of beta graduation critera

Reply:
done. added the paths of the tests.